File: 0200-number-of-islands/0200-number-of-islands.cpp
class Solution {
public:
    vector<pair<int,int>> dirs = {{0,1}, {0,-1}, {1,0}, {-1,0}};
    int dfs(int i, int j, vector<vector<char>>& grid){
        if (i < 0 || i >= grid.size() || j < 0 || j >= grid[0].size() || grid[i][j]=='0' || grid[i][j]=='2'){
            return 0;
        }
        
        grid[i][j] = '2';
        for (auto dir : dirs){
            dfs(i+dir.first, j+dir.second, grid);
        }
        return 1;
    }
    
    int numIslands(vector<vector<char>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        
        int result = 0;
        for (int i=0; i<m; i++){
            for (int j=0; j<n; j++){
                if (grid[i][j] == '1'){
                    result += dfs(i,j,grid);                
                }
            }
        }
        return result;
    }
};

File: 0200-number-of-islands/0200-number-of-islands.py
class Solution:
    def __init__(self):
        self.dirs = [(1,0), (-1,0), (0, 1), (0, -1)]
    def numIslands(self, grid: List[List[str]]) -> int:
        ans = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                ans += self.helper(i, j, grid)
        return ans

    def helper(self, i, j, grid):
        if i < 0 or i >= len(grid) or j < 0 or j >= len(grid[0]):
            return 0
        spot = grid[i][j]
        if spot != "1":
            return 0
        grid[i][j] = "2"
        for direction in self.dirs:
            self.helper(i + direction[0], j + direction[1], grid)
        return 1
        

